Revit_Core_Engine/Query/FacadeCurtainPanels.cs

```diff
@@ -0,0 +1,108 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using Autodesk.Revit.DB;
+using BH.oM.Adapters.Revit.Settings;
+using BH.oM.Base;
+using BH.oM.Dimensional;
+using BH.oM.Geometry;
+using BH.oM.Facade.Elements;
+using BH.Engine.Facade;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        public static List<oM.Facade.Elements.Opening> FacadeCurtainPanels(this CurtainGrid curtainGrid, Document document, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
+        {
+            if (curtainGrid == null)
+                return null;
+
+            List<Element> panels = curtainGrid.GetPanelIds().Select(x => document.GetElement(x)).ToList();
+            List<CurtainCell> cells = curtainGrid.GetCurtainCells().ToList();
+
+            if (panels.Count != cells.Count)
+                return null;
+
+            List<IElement1D> cwMullions = curtainGrid.CurtainWallMullions(document, settings, refObjects).Select(x => x as IElement1D).ToList();
+
+            List<oM.Facade.Elements.Opening> result = new List<oM.Facade.Elements.Opening>();
+
+            for (int i = 0; i < panels.Count; i++)
+            {
+                FamilyInstance panel = panels[i] as FamilyInstance;
+                List<PolyCurve> pcs = new List<PolyCurve>();
+                if (panel == null)
+                    continue;
+                try
+                { 
+                    pcs = cells[i].CurveLoops.FromRevit(); 
+                }
+                catch  // This catches when CurveLoops throws an exception due to the cell having no loops, meaning in Revit it exists in the database but is no longer on the CurtainWall with any corresponding Curves
+                { 
+                    continue; 
+                }
+
+                foreach (PolyCurve pc in pcs)
+                {
+                    BH.oM.Facade.Elements.Opening bHoMOpening = null;
+                    // If panel is a basic wall, the panel is not the actual element, it is an empty panel that hosts the
+                    // actual element, so we assign it a null construction and separately return the wall
+                    if (panel is Autodesk.Revit.DB.Panel elemPanel && document.GetElement(elemPanel.FindHostPanel()) is Wall hostElement)
```

Comment:
Apparently, you are using pattern matching that has been introduced in C#7. Since we are still trying to support VS2015, it is preferable to avoid anything newer than C#6 - would you be OK with declaring the variables within the brackets instead of doing it on `is` checks? I know this will be an extra pain, but it would be a shame to lose the support for older (and free) versions of VS due to that...

Reply:
Yea, fair enough, I'll change that. Pattern matching is so neat though! :)